File: sdk-typescript/src/async-client.ts
/**
 * HandoffRail SDK — Asynchronous HTTP client.
 *
 * Uses the native `fetch` API (available in Node.js 18+).
 *
 * Usage:
 * ```typescript
 * import { AsyncHandoffRailClient, PacketBuilder } from 'handoffrail-sdk';
 *
 * const client = new AsyncHandoffRailClient({
 *   baseUrl: 'http://localhost:8080/api/v1',
 *   apiKey: 'sk-...',
 * });
 *
 * const packet = await client.createPacket(
 *   new PacketBuilder()
 *     .to('billing-01', 'BillingBot')
 *     .from('sales-01', 'SalesBot')
 *     .withSummary('Customer wants Business tier')
 *     .build()
 * );
 *
 * await client.close();
 * ```
 *
 * @module
 */

import type {
  PacketCreate,
  PacketResponse,
  PacketListResponse,
  AuditLogResponse,
  PacketUpdate,
  PacketHistoryResponse,
  ChainHandoffRequest,
  WebhookCreate,
  WebhookResponse,
  ListPacketsOptions,
  ListAuditOptions,
  ClaimPacketOptions,
  HitlRespondOptions,
  RegisterWebhookOptions,
  WebhookDelivery,
  BatchCreateResponse,
  BatchClaimOptions,
  BatchClaimResponse,
  BatchCompleteResponse,
  SearchOptions,
} from './models';

import {
  serializePacketCreate,
  serializePacketUpdate,
  serializeChainHandoffRequest,
} from './models';

import {
  HandoffRailError,
  AuthenticationError,
  NotFoundError,
  ValidationError,
  ConflictError,
  RateLimitError,
  ServerError,
  ConnectionError,
} from './errors';

// ── Configuration ────────────────────────────────────────────────────────

/** Configuration options for the async HandoffRail client. */
export interface AsyncHandoffRailClientOptions {
  /** Base URL of the HandoffRail API (e.g. `http://localhost:8080/api/v1`). */
  baseUrl: string;
  /** API key for authentication (sent as `X-API-Key` header). */
  apiKey: string;
  /** Request timeout in milliseconds (default 30000). */
  timeout?: number;
  /** Maximum number of retries on transient errors (default 3). */
  maxRetries?: number;
  /** Base delay in milliseconds between retries (uses exponential backoff, default 500). */
  retryDelay?: number;
}

// ── Helper: sleep ────────────────────────────────────────────────────────

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

// ── Client ───────────────────────────────────────────────────────────────

/**
 * Asynchronous HTTP client for the HandoffRail API.
 *
 * @remarks
 * All methods return Promises. For blocking/synchronous usage, see
 * {@link HandoffRailClient}.
 */
export class AsyncHandoffRailClient {
  public readonly baseUrl: string;
  public readonly apiKey: string;
  public readonly timeout: number;
  public readonly maxRetries: number;
  public readonly retryDelay: number;
  private _abortController: AbortController | null = null;

  constructor(options: AsyncHandoffRailClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, '');
    this.apiKey = options.apiKey;
    this.timeout = options.timeout ?? 30_000;
    this.maxRetries = options.maxRetries ?? 3;
    this.retryDelay = options.retryDelay ?? 500;
  }

  // ── Public lifecycle ──────────────────────────────────────────────────

  /**
   * Close any pending connections.
   */
  async close(): Promise<void> {
    if (this._abortController) {
      this._abortController.abort();
      this._abortController = null;
    }
  }

  // ── Internal helpers ──────────────────────────────────────────────────

  /**
   * Build default headers for every request.
   */
  private _headers(): Record<string, string> {
    return {
      'X-API-Key': this.apiKey,
      'Content-Type': 'application/json',
    };
  }

  /**
   * Parse the JSON response body.
   */
  private async _parseBody(
    response: Response,
  ): Promise<Record<string, unknown> | Record<string, unknown>[] | null> {
    const text = await response.text();
    if (!text) return null;
    try {
      const parsed = JSON.parse(text);
      if (Array.isArray(parsed)) return parsed;
      if (typeof parsed === 'object' && parsed !== null) return parsed as Record<string, unknown>;
      return null;
    } catch {
      return { raw: text };
    }
  }

  /**
   * Handle the HTTP response and map status codes to SDK errors.
   */
  private async _handleResponse(
    response: Response,
    url: string,
  ): Promise<Record<string, unknown> | Record<string, unknown>[]> {
    const body = await this._parseBody(response);
    const responseBody = (body as Record<string, unknown>) ?? {};

    if (response.status === 401) {
      throw new AuthenticationError(
        (responseBody.detail as string) ?? 'Authentication failed',
        { responseBody },
      );
    }
    if (response.status === 404 || response.status === 410) {
      throw new NotFoundError(
        (responseBody.detail as string) ?? 'Resource not found',
        { statusCode: response.status, responseBody, resourceId: url },
      );
    }
    if (response.status === 400) {
      throw new ValidationError(
        (responseBody.detail as string) ?? 'Validation error',
        { field: responseBody.field as string | undefined, responseBody },
      );
    }
    if (response.status === 409) {
      const detail = responseBody.detail ?? responseBody;
      throw new ConflictError(
        typeof detail === 'string' ? detail : 'Resource conflict',
        { responseBody },
      );
    }
    if (response.status === 429) {
      const retryAfter = response.headers.get('Retry-After');
      throw new RateLimitError('Rate limit exceeded', {
        retryAfter: retryAfter ? parseInt(retryAfter, 10) : undefined,
        responseBody,
      });
    }
    if (response.status >= 500) {
      throw new ServerError(`Server error: ${response.status}`, {
        statusCode: response.status,
        responseBody,
      });
    }
    if (response.status === 204) {
      return {};
    }

    // Success
    return body ?? {};
  }

  /**
   * Make an HTTP request with retry logic and error mapping.
   */
  protected async _request(
    method: string,
    path: string,
    options?: {
      jsonData?: Record<string, unknown>;
      params?: Record<string, string | number | undefined>;
    },
  ): Promise<Record<string, unknown> | Record<string, unknown>[]> {
    const url = new URL(`${this.baseUrl}${path}`);

    if (options?.params) {
      for (const [key, value] of Object.entries(options.params)) {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }

    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      let timeoutId: ReturnType<typeof setTimeout> | null = null;
      try {
        this._abortController = new AbortController();
        timeoutId = setTimeout(() => this._abortController?.abort(), this.timeout);

        const fetchOptions: RequestInit & { signal: AbortSignal } = {
          method,
          headers: {
            'X-API-Key': this.apiKey,
            ...(options?.jsonData ? { 'Content-Type': 'application/json' } : {}),
          },
          signal: this._abortController.signal,
        };

        if (options?.jsonData !== undefined) {
          fetchOptions.body = JSON.stringify(options.jsonData);
        }

        const response = await fetch(url.toString(), fetchOptions);
        const result = await this._handleResponse(response, url.toString());
        return result;
      } catch (err) {
        // Rethrow HandoffRailErrors directly (e.g., 4xx/5xx from _handleResponse)
        if (err instanceof HandoffRailError) {
          throw err;
        }

        lastError = err as Error;

        // Don't retry on abort (timeout)
        if (err instanceof DOMException && err.name === 'AbortError') {
          throw new ConnectionError('Request timed out', { originalError: err as Error });
        }

        if (attempt < this.maxRetries) {
          const backoff = this.retryDelay * Math.pow(2, attempt);
          await sleep(backoff);
          continue;
        }
      } finally {
        // Always clear timeout and controller — prevents open handles in tests
        if (timeoutId !== null) {
          clearTimeout(timeoutId);
        }
        this._abortController = null;
      }
    }

    throw new ConnectionError(
      `Unable to connect to HandoffRail server after ${this.maxRetries + 1} attempts`,
      { originalError: lastError ?? undefined },
    );
  }

  // ── Packet CRUD ────────────────────────────────────────────────────────

  /**
   * Create a new handoff packet.
   */
  async createPacket(packet: PacketCreate): Promise<PacketResponse> {
    const data = await this._request('POST', '/packets', {
      jsonData: serializePacketCreate(packet) as Record<string, unknown>,
    });
    return data as unknown as PacketResponse;
  }

  /**
   * Get a single packet by ID.
   */
  async getPacket(packetId: string): Promise<PacketResponse> {
    const data = await this._request('GET', `/packets/${encodeURIComponent(packetId)}`);
    return data as unknown as PacketResponse;
  }

  /**
   * List packets with filtering and pagination.
   */
  async listPackets(options?: ListPacketsOptions): Promise<PacketListResponse> {
    const params: Record<string, string | number | undefined> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.status) params.status = options.status;
    if (options?.source_agent) params.source_agent = options.source_agent;
    if (options?.target_agent) params.target_agent = options.target_agent;
    if (options?.tags) params.tags = options.tags;
    if (options?.priority) params.priority = options.priority;
    if (options?.created_after) params.created_after = options.created_after;
    if (options?.created_before) params.created_before = options.created_before;
    if (options?.cursor) params.cursor = options.cursor;

    const data = await this._request('GET', '/packets', { params });
    return data as unknown as PacketListResponse;
  }

  /**
   * List structured audit log entries.
   */
  async listAuditLog(options?: ListAuditOptions): Promise<AuditLogResponse> {
    const params: Record<string, string | number | undefined> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.actor) params.actor = options.actor;
    if (options?.action) params.action = options.action;
    if (options?.packet_id) params.packet_id = options.packet_id;
    if (options?.created_after) params.created_after = options.created_after;
    if (options?.created_before) params.created_before = options.created_before;

    const data = await this._request('GET', '/audit', { params });
    return data as unknown as AuditLogResponse;
  }

  /**
   * Claim a packet for processing.
   */
  async claimPacket(
    packetId: string,
    options: ClaimPacketOptions,
  ): Promise<PacketResponse> {
    const data = await this._request(
      'POST',
      `/packets/${encodeURIComponent(packetId)}/claim`,
      {
        jsonData: {
          agent_id: options.agent_id,
          agent_name: options.agent_name,
          framework: options.framework,
        },
      },
    );
    return data as unknown as PacketResponse;
  }

  /**
   * Partially update a packet.
   */
  async updatePacket(
    packetId: string,
    update: PacketUpdate,
  ): Promise<PacketResponse> {
    const data = await this._request(
      'PATCH',
      `/packets/${encodeURIComponent(packetId)}`,
      { jsonData: serializePacketUpdate(update) as Record<string, unknown> },
    );
    return data as unknown as PacketResponse;
  }

  /**
   * Mark a packet as completed.
   *
   * Convenience method that sets status to `completed`.
   */
  async completePacket(packetId: string): Promise<PacketResponse> {
    return this.updatePacket(packetId, { status: 'completed' });
  }

  /**
   * Soft-delete a packet (marks as expired).
   */
  async deletePacket(packetId: string): Promise<void> {
    await this._request('DELETE', `/packets/${encodeURIComponent(packetId)}`);
  }

  /**
   * Submit a human response to a HITL checkpoint.
   */
  async respondToHitl(
    packetId: string,
    options: HitlRespondOptions,
  ): Promise<PacketResponse> {
    const data = await this._request(
      'POST',
      `/packets/${encodeURIComponent(packetId)}/respond`,
      {
        jsonData: {
          response: options.response,
          responded_by: options.responded_by,
          notes: options.notes,
        },
      },
    );
    return data as unknown as PacketResponse;
  }

  /**
   * Get packets awaiting human review.
   */
  async listAwaitingHuman(
    options?: { limit?: number; offset?: number },
  ): Promise<PacketListResponse> {
    const params: Record<string, string | number | undefined> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    const data = await this._request('GET', '/packets/awaiting', { params });
    return data as unknown as PacketListResponse;
  }

  /**
   * Get the event history for a packet.
   */
  async getPacketHistory(packetId: string): Promise<PacketHistoryResponse> {
    const data = await this._request(
      'GET',
      `/packets/${encodeURIComponent(packetId)}/history`,
    );
    return data as unknown as PacketHistoryResponse;
  }

  /**
   * Create a chained follow-up packet.
   */
  async chainPacket(
    parentPacketId: string,
    request: ChainHandoffRequest,
  ): Promise<PacketResponse> {
    const data = await this._request(
      'POST',
      `/packets/${encodeURIComponent(parentPacketId)}/chain`,
      { jsonData: serializeChainHandoffRequest(request) as Record<string, unknown> },
    );
    return data as unknown as PacketResponse;
  }

  // ── Webhook CRUD ───────────────────────────────────────────────────────

  /**
   * Register a new webhook.
   */
  async registerWebhook(options: RegisterWebhookOptions): Promise<WebhookResponse> {
    const data = await this._request('POST', '/hooks', {
      jsonData: {
        url: options.url,
        events: options.events ?? [
          'packet.created',
          'packet.claimed',
          'packet.completed',
          'packet.failed',
        ],
        secret: options.secret,
      },
    });
    return data as unknown as WebhookResponse;
  }

  /**
   * List all webhooks for the authenticated tenant.
   */
  async listWebhooks(): Promise<WebhookResponse[]> {
    const data = await this._request('GET', '/hooks');
    if (Array.isArray(data)) {
      return data.map((w) => w as unknown as WebhookResponse);
    }
    return [];
  }

  /**
   * Deactivate (soft-delete) a webhook.
   */
  async deleteWebhook(webhookId: string): Promise<void> {
    await this._request('DELETE', `/hooks/${encodeURIComponent(webhookId)}`);
  }

  /**
   * List delivery history for one webhook.
   */
  async listWebhookDeliveries(
    webhookId: string,
    options?: { status?: string; limit?: number; offset?: number },
  ): Promise<WebhookDelivery[]> {
    const params: Record<string, string | number | undefined> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.status) params.status = options.status;

    const data = await this._request('GET', `/hooks/${encodeURIComponent(webhookId)}/deliveries`, { params });
    return Array.isArray(data) ? (data as unknown as WebhookDelivery[]) : [];
  }

  // ── Batch Operations ───────────────────────────────────────────────────

  /**
   * Create multiple packets in a single request (max 50).
   */
  async batchCreatePackets(packets: PacketCreate[]): Promise<BatchCreateResponse> {
    const payload = { packets: packets.map(p => serializePacketCreate(p)) };
    const data = await this._request('POST', '/packets/batch', { jsonData: payload });
    return data as unknown as BatchCreateResponse;
  }

  /**
   * Claim multiple packets in a single request.
   */
  async batchClaimPackets(packetIds: string[], options: BatchClaimOptions): Promise<BatchClaimResponse> {
    const payload = {
      packet_ids: packetIds,
      agent_id: options.agentId,
      agent_name: options.agentName,
      ...(options.framework && { framework: options.framework }),
    };
    const data = await this._request('POST', '/packets/batch/claim', { jsonData: payload });
    return data as unknown as BatchClaimResponse;
  }

  /**
   * Complete multiple packets in a single request.
   */
  async batchCompletePackets(packetIds: string[]): Promise<BatchCompleteResponse> {
    const payload = { packet_ids: packetIds };
    const data = await this._request('POST', '/packets/batch/complete', { jsonData: payload });
    return data as unknown as BatchCompleteResponse;
  }

  // ── Search ─────────────────────────────────────────────────────────────

  /**
   * Full-text search across packet summaries and context.
   */
  async searchPackets(query: string, options?: SearchOptions): Promise<PacketListResponse> {
    const params: Record<string, string | number | undefined> = { q: query };
    if (options) {
      if (options.limit !== undefined) params.limit = options.limit;
      if (options.offset !== undefined) params.offset = options.offset;
      if (options.status) params.status = options.status;
      if (options.priority) params.priority = options.priority;
    }
    const data = await this._request('GET', '/packets/search', { params });
    return data as unknown as PacketListResponse;
  }
}

File: sdk/src/handoffrail/sdk/async_client.py
"""HandoffRail SDK — Asynchronous HTTP client.

Usage::

    from handoffrail.sdk import AsyncHandoffRailClient

    async with AsyncHandoffRailClient(base_url="...", api_key="hr_...") as client:
        packet = await client.create_packet(...)
        claimed = await client.claim_packet(packet.id, agent_id="billing-01", agent_name="BillingBot")
"""

from __future__ import annotations

import asyncio
from typing import Any
from uuid import UUID

import httpx

from handoffrail.sdk.exceptions import (
    AuthenticationError,
    ConnectionError,
    HandoffRailError,
    NotFoundError,
    RateLimitError,
    ServerError,
    ValidationError,
)
from handoffrail.sdk.models import (
    BatchClaimRequest,
    BatchClaimResponse,
    BatchCompleteRequest,
    BatchCompleteResponse,
    BatchCreateResponse,
    ChainHandoffRequest,
    HitlRespondRequest,
    PacketClaim,
    PacketCreate,
    PacketHistoryResponse,
    PacketListResponse,
    PacketResponse,
    PacketUpdate,
    SearchOptions,
    WebhookCreate,
    WebhookResponse,
)


class AsyncHandoffRailClient:
    """Asynchronous HTTP client for the HandoffRail API.

    Args:
        base_url: Base URL of the HandoffRail API (e.g. ``http://localhost:8080/api/v1``).
        api_key: API key for authentication (sent as ``X-API-Key`` header).
        timeout: Request timeout in seconds (default 30).
        max_retries: Maximum number of retries on transient errors (default 3).
        retry_delay: Base delay in seconds between retries (uses exponential backoff).
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float = 30.0,
        max_retries: int = 3,
        retry_delay: float = 0.5,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            headers={"X-API-Key": self.api_key},
            timeout=self.timeout,
        )

    # ── Context manager ────────────────────────────────────────────────────

    async def __aenter__(self) -> AsyncHandoffRailClient:
        return self

    async def __aexit__(self, exc_type: type | None, exc_val: BaseException | None, exc_tb: Any | None) -> None:
        await self.close()

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._client.aclose()

    # ── Internal helpers ───────────────────────────────────────────────────

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json_data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Make an async HTTP request with retry logic and error mapping."""
        last_exception: Exception | None = None

        for attempt in range(self.max_retries + 1):
            try:
                response = await self._client.request(
                    method,
                    path,
                    json=json_data,
                    params=params,
                )

                if response.status_code == 401:
                    raise AuthenticationError(
                        response.json().get("detail", "Authentication failed"),
                        status_code=401,
                        response_body=response.json(),
                    )
                if response.status_code == 404:
                    raise NotFoundError(
                        response.json().get("detail", "Resource not found"),
                        status_code=404,
                        response_body=response.json(),
                    )
                if response.status_code == 410:
                    raise NotFoundError(
                        response.json().get("detail", "Resource gone"),
                        resource_id=path,
                        status_code=410,
                        response_body=response.json(),
                    )
                if response.status_code == 400:
                    body = response.json()
                    raise ValidationError(
                        body.get("detail", "Validation error"),
                        field=body.get("field"),
                        status_code=400,
                        response_body=body,
                    )
                if response.status_code == 409:
                    body = response.json()
                    detail = body.get("detail", body) if isinstance(body, dict) else str(body)
                    raise ValidationError(
                        str(detail) if not isinstance(detail, str) else detail,
                        status_code=409,
                        response_body=body,
                    )
                if response.status_code == 429:
                    retry_after = response.headers.get("Retry-After")
                    raise RateLimitError(
                        "Rate limit exceeded",
                        retry_after=int(retry_after) if retry_after else None,
                        status_code=429,
                        response_body=response.json() if response.content else {},
                    )
                if response.status_code >= 500:
                    raise ServerError(
                        f"Server error: {response.status_code}",
                        status_code=response.status_code,
                        response_body=response.json() if response.content else {},
                    )
                if response.status_code == 204:
                    return {}

                response.raise_for_status()
                return response.json()

            except (httpx.ConnectError, httpx.TimeoutException) as exc:
                last_exception = exc
                if attempt < self.max_retries:
                    backoff = self.retry_delay * (2 ** attempt)
                    await asyncio.sleep(backoff)
                    continue
                raise ConnectionError(
                    f"Unable to connect to HandoffRail server after {self.max_retries + 1} attempts",
                    original_error=exc,
                ) from exc

            except (AuthenticationError, NotFoundError, ValidationError, RateLimitError, ServerError):
                raise

            except httpx.HTTPStatusError as exc:
                raise ServerError(
                    f"Unexpected HTTP error: {exc.response.status_code}",
                    status_code=exc.response.status_code,
                    response_body=exc.response.json() if exc.response.content else {},
                ) from exc

            except Exception as exc:
                if isinstance(exc, HandoffRailError):
                    raise
                last_exception = exc
                if attempt < self.max_retries:
                    backoff = self.retry_delay * (2 ** attempt)
                    await asyncio.sleep(backoff)
                    continue
                raise ConnectionError(
                    "Unexpected error communicating with HandoffRail server",
                    original_error=exc,
                ) from exc

        raise ConnectionError("Exhausted retries", original_error=last_exception)

    # ── Packet CRUD ────────────────────────────────────────────────────────

    async def create_packet(self, packet: PacketCreate) -> PacketResponse:
        """Create a new handoff packet."""
        data = await self._request("POST", "/packets", json_data=packet.to_dict())
        return PacketResponse.from_dict(data)

    async def get_packet(self, packet_id: str | UUID) -> PacketResponse:
        """Get a single packet by ID."""
        data = await self._request("GET", f"/packets/{packet_id}")
        return PacketResponse.from_dict(data)

    async def list_packets(
        self,
        *,
        status: str | None = None,
        source_agent: str | None = None,
        target_agent: str | None = None,
        tags: str | None = None,
        priority: str | None = None,
        created_after: str | None = None,
        created_before: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> PacketListResponse:
        """List packets with filtering and pagination."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status
        if source_agent:
            params["source_agent"] = source_agent
        if target_agent:
            params["target_agent"] = target_agent
        if tags:
            params["tags"] = tags
        if priority:
            params["priority"] = priority
        if created_after:
            params["created_after"] = created_after
        if created_before:
            params["created_before"] = created_before

        data = await self._request("GET", "/packets", params=params)
        return PacketListResponse.from_dict(data)

    async def claim_packet(
        self,
        packet_id: str | UUID,
        *,
        agent_id: str,
        agent_name: str,
        framework: str | None = None,
    ) -> PacketResponse:
        """Claim a packet for processing."""
        claim = PacketClaim(agent_id=agent_id, agent_name=agent_name, framework=framework)
        data = await self._request("POST", f"/packets/{packet_id}/claim", json_data=claim.to_dict())
        return PacketResponse.from_dict(data)

    async def update_packet(self, packet_id: str | UUID, update: PacketUpdate) -> PacketResponse:
        """Partially update a packet."""
        data = await self._request("PATCH", f"/packets/{packet_id}", json_data=update.to_dict())
        return PacketResponse.from_dict(data)

    async def complete_packet(self, packet_id: str | UUID) -> PacketResponse:
        """Mark a packet as completed."""
        update = PacketUpdate(status="completed")
        return await self.update_packet(packet_id, update)

    async def delete_packet(self, packet_id: str | UUID) -> None:
        """Soft-delete a packet (marks as expired)."""
        await self._request("DELETE", f"/packets/{packet_id}")

    async def respond_to_hitl(
        self,
        packet_id: str | UUID,
        *,
        response: str,
        responded_by: str,
        notes: str | None = None,
    ) -> PacketResponse:
        """Submit a human response to a HITL checkpoint."""
        payload = HitlRespondRequest(response=response, responded_by=responded_by, notes=notes)
        data = await self._request("POST", f"/packets/{packet_id}/respond", json_data=payload.to_dict())
        return PacketResponse.from_dict(data)

    async def get_awaiting(
        self,
        *,
        limit: int = 50,
        offset: int = 0,
    ) -> PacketListResponse:
        """Get packets awaiting human review."""
        data = await self._request("GET", "/packets/awaiting", params={"limit": limit, "offset": offset})
        return PacketListResponse.from_dict(data)

    async def get_history(self, packet_id: str | UUID) -> PacketHistoryResponse:
        """Get the event history for a packet."""
        data = await self._request("GET", f"/packets/{packet_id}/history")
        return PacketHistoryResponse.from_dict(data)

    async def chain_handoff(self, parent_packet_id: str | UUID, request: ChainHandoffRequest) -> PacketResponse:
        """Create a chained follow-up packet."""
        data = await self._request("POST", f"/packets/{parent_packet_id}/chain", json_data=request.to_dict())
        return PacketResponse.from_dict(data)

    # ── Webhook CRUD ───────────────────────────────────────────────────────

    async def register_webhook(
        self,
        *,
        url: str,
        events: list[str] | None = None,
        secret: str,
    ) -> WebhookResponse:
        """Register a new webhook."""
        webhook = WebhookCreate(url=url, events=events or ["packet.created", "packet.claimed", "packet.completed", "packet.failed"], secret=secret)
        data = await self._request("POST", "/hooks", json_data=webhook.to_dict())
        return WebhookResponse.from_dict(data)

    async def list_webhooks(self) -> list[WebhookResponse]:
        """List all webhooks for the authenticated tenant."""
        data = await self._request("GET", "/hooks")
        return [WebhookResponse.from_dict(w) for w in data]

    async def delete_webhook(self, webhook_id: str) -> None:
        """Deactivate (soft-delete) a webhook."""
        await self._request("DELETE", f"/hooks/{webhook_id}")

    # ── Batch Operations ───────────────────────────────────────────────────────

    async def batch_create_packets(self, packets: list[PacketCreate]) -> BatchCreateResponse:
        """Create multiple packets in a single request.

        Args:
            packets: List of packet create payloads (max 50).

        Returns:
            Response with created packets and any errors.
        """
        payload = {"packets": [p.to_dict() for p in packets]}
        data = await self._request("POST", "/packets/batch", json_data=payload)
        return BatchCreateResponse.from_dict(data)

    async def batch_claim_packets(
        self,
        packet_ids: list[str],
        agent_id: str,
        agent_name: str,
        framework: str | None = None,
    ) -> BatchClaimResponse:
        """Claim multiple packets in a single request.

        Args:
            packet_ids: List of packet UUIDs to claim.
            agent_id: The claiming agent's ID.
            agent_name: The claiming agent's name.
            framework: Optional agent framework identifier.

        Returns:
            Response with claimed packets and any errors.
        """
        payload = BatchClaimRequest(
            packet_ids=packet_ids,
            agent_id=agent_id,
            agent_name=agent_name,
            framework=framework,
        )
        data = await self._request("POST", "/packets/batch/claim", json_data=payload.to_dict())
        return BatchClaimResponse.from_dict(data)

    async def batch_complete_packets(self, packet_ids: list[str]) -> BatchCompleteResponse:
        """Complete multiple packets in a single request.

        Args:
            packet_ids: List of packet UUIDs to complete.

        Returns:
            Response with completed packets and any errors.
        """
        payload = BatchCompleteRequest(packet_ids=packet_ids)
        data = await self._request("POST", "/packets/batch/complete", json_data=payload.to_dict())
        return BatchCompleteResponse.from_dict(data)

    # ── Search ─────────────────────────────────────────────────────────────────

    async def search_packets(self, query: str, options: SearchOptions | None = None) -> PacketListResponse:
        """Full-text search across packet summaries and context.

        Args:
            query: Search query string (min 2 characters).
            options: Optional filters (limit, offset, status, priority).

        Returns:
            List response with matching packets ranked by relevance.
        """
        params: dict[str, Any] = {"q": query}
        if options:
            params.update(options.to_params())
        data = await self._request("GET", "/packets/search", params=params)
        return PacketListResponse.from_dict(data)
